### PR TITLE
Add username copying to password manager.

### PR DESCRIPTION
**Pull request by aartaka**
This adds username copying to KeePassXC and Password Store interfaces, together with a generic to implement for other interfaces. A step towards #1138.

EDIT: `command` is installed on my machine, but isn't installed in development environment xD
EDIT: Mention #1138.

### CODE REVIEW

**Comment by aartaka** [comment]
Any `pass` users to test it?

**Comment by aartaka** [comment]
Important pointers to learn from:
[Browserpass conventions for passwords and usernames](https://github.com/browserpass/browserpass-extension#organizing-password-store)
[auth-source-pass](https://github.com/DamienCassou/auth-source-pass)
[passff](https://github.com/passff/passff#readme)

**Comment by edgar-vincent** [comment]
Just to keep a note of this: in its current state, `copy-username` copies the entire line containing the username, i.e `login:  mylogin` instead of `mylogin`.

File: libraries/password-manager/password-security.lisp [hunk 1]
```diff
@@ -29,4 +29,8 @@
                         "-a" password-name "-s" service "-g")
                   :error-output '(:string :stripped t)))))))
 
+(defmethod clip-username ((password-interface security-interface) &key password-name service)
+  (declare (ignore password-name service))
+  (error "Username clipping is not supported by security interface."))
```
Comment (Ambrevar) [inline libraries/password-manager/password-security.lisp:34]: Ambiguous use of "security" here.
Maybe `by the 'security' interface`?

File: libraries/password-manager/password.lisp [hunk 2]
```diff
@@ -23,9 +23,15 @@
 (defgeneric clip-password (password-interface &key password-name service)
   (:documentation "Retrieve specific password by name."))
 
+(eval-when (:compile-toplevel :load-toplevel :execute)
+  (export 'clip-username))
```
Comment (Ambrevar) [inline libraries/password-manager/password.lisp:27]: We should probably define the `export-always` macro locally at this point :)

File: libraries/password-manager/password-pass.lisp [hunk 3]
```diff
@@ -39,13 +39,26 @@
...
     :output '(:string :stripped t)))
 
+(defmethod clip-username ((password-interface password-store-interface) &key password-name service)
+  (declare (ignore service))
+  (let ((username-line (1+ (position "[Ll]ogin|[Uu]ser(name)?:"
+                                     (ppcre:split "\\n"
+                                                  (execute password-interface
+                                                      (list "show" password-name)
+                                                    :output '(:string :stripped t)))
+                                     :test #'ppcre:all-matches))))
```
Comment (Ambrevar) [inline libraries/password-manager/password-pass.lisp:49]: username-line may be NIL, `1+` will raise a condition then.
Furthermore, the rest of the body should not be run if `username-line` is NIL.

File: libraries/password-manager/password-pass.lisp [hunk 4]
```diff
@@ -39,13 +39,26 @@
   (execute password-interface (list "show" "--clip" password-name)
     :output '(:string :stripped t)))
 
+(defmethod clip-username ((password-interface password-store-interface) &key password-name service)
+  (declare (ignore service))
+  (let ((username-line (1+ (position "[Ll]ogin|[Uu]ser(name)?:"
```
Comment (Ambrevar) [inline libraries/password-manager/password-pass.lisp:44]: The username maybe not be prefixed with anything at all, and just be on the second line by convention.  I suggest you default to `2` if there is no result.  Bear in mind there may be no second line.

File: libraries/password-manager/password-pass.lisp [hunk 5]
```diff
@@ -39,13 +39,26 @@
...
+                                     (ppcre:split "\\n"
+                                                  (execute password-interface
+                                                      (list "show" password-name)
+                                                    :output '(:string :stripped t)))
+                                     :test #'ppcre:all-matches))))
+    (execute password-interface (list "show" (format nil "--clip=~d" username-line) password-name)
+      :output '(:string :stripped t))))
```
Comment (Ambrevar) [inline libraries/password-manager/password-pass.lisp:51]: You need to strip the matching "[Ll]ogin|[Uu]ser(name)?:" as reported by @edgar-vincent.

**Comment by Ambrevar** [comment]
Great addition, thanks!

What about going even more generic and add a command to copy _any_ field?
In the case of `pass`, it could simply parse the entry and offer each line as a suggestion in the prompt buffer.
If a line has `:` or `=`, we could offer to copy only the right-hand side.

File: libraries/password-manager/password-pass.lisp (hunk 4 above)
Comment (edgar-vincent) [reply libraries/password-manager/password-pass.lisp:44]: I don't think having an unprefixed field is encouraged: what if an entry doesn't have a user associated to it but does have a port or a URL, stored on the second line? Moreover,  a password could contain a `:` or a `=`. I think we should stick to a specific organisational style, like what `auth-source-pass` for Emacs uses.

**Comment by Ambrevar** [comment]
Shall we fix and merge this before the next release? @aartaka Do you have time to fix it?  Let me know if you need help! :)

**Comment by aartaka** [comment]
I'll be glad to have help there, because I'm not a `pass` user and I barely understand the conventions behind it :)

All in all, this seems to be too draft-ish to merge just now.

**Comment by edgar-vincent** [comment]
The conventions for storing additional data in `pass` entries are not standardised. I suggest we follow the following organisational rules:

- the first line is the password except if any of the subsequent lines is prefixed with `password: `, `pass: ` or `secret: ` (case-insensitive)
- the username entry should be prefixed with `login: `, `username: `, or `user: ` (case-insensitive)
- additional fields, such as `port: `, are stored in subsequent lines and are always prefixed by a case-insensitive `keyword: `.
- an entry's name (eg: `My_websites/example.com`) should contain the full domain name for a given website
- if there are two or more accounts for a given domain name, then they can be specified in the entry's path: 
  * either `My_websites/example.com/myuser`
  * or `My_websites/myuser@example.com`
- a port number can be specified after a colon in a given entry's name: `My_websites/example.com:443`

This should provide acceptable compatibility with https://github.com/DamienCassou/auth-source-pass https://github.com/passff/passff#readme and https://github.com/browserpass/browserpass-extension#organizing-password-store

**Comment by aartaka** [comment]
@edgar-vincent, thanks for this summary!

@Ambrevar, feel free to take this issue over -- I don't have enough time to do it well :(

**Comment by Ambrevar** [comment]
I've just pushed some commits which should address most of the comments in this thread.

@edgar-vincent Can you test it and report if this cuts it for you?  Thanks in advance :)

> the first line is the password except if any of the subsequent lines is prefixed with password: , pass: or secret: (case-insensitive)

I haven't implemented this because it adds some complexity and I'm not sure this is really in use.  `password-store` home page really insists that the password ought to be on the first line.

About the port: I don't have a use case, but is Nyxt missing something here?

File: libraries/password-manager/password-pass.lisp [hunk 6]
```diff
@@ -39,13 +39,58 @@
...
+                            (when (= 2 (length entry))
+                              entry)))
+                        ;; Skip first line to ignore password:
+                        (rest lines)))))
+
+(defvar *username-keys* '("login" "user" "username")
+  "A list of string keys used to find the `pass' username in `clip-username'.")
+
+(defmethod clip-username ((password-interface password-store-interface) &key password-name service)
+  "Save the multiline entry that's prefixed with on of the `*username-keys*' to clipboard.
```
Comment (edgar-vincent) [inline libraries/password-manager/password-pass.lisp:71]: s/on of/one of

**Comment by edgar-vincent** [comment]
> I've just pushed some commits which should address most of the comments in this thread.
> 

Thank you *very* much for this! It seems to work wonderfully.

> > the first line is the password except if any of the subsequent lines is prefixed with password: , pass: or secret: (case-insensitive)
> 
> I haven't implemented this because it adds some complexity and I'm not sure this is really in use. `password-store` home page really insists that the password ought to be on the first line.

I agree. I just mentioned it for the sake of compatibility with other add-ons, but your choice makes a lot of sense.

> About the port: I don't have a use case, but is Nyxt missing something here?

If might be useful if, for example, one has service A running on https://localhost:123 and service B on https://localhost:456, both services requiring different credentials.

**Comment by Ambrevar** [comment]
@edgar-vincent Thanks for your feedback.

I've merged this as 2f9f500ec60f0001530a8dd5bd841daf5fa28ac5 since it's working and has been hanging for too long.

About the port question: for now Nyxt makes no attempt at guessing the right password entry based on the current buffer, the URL / port has no impact on the UX at the moment.  

If I'm mistaken, can you detail a precise recipe where parsing the port would be useful for the Nyxt UI?

**Comment by edgar-vincent** [comment]
Thanks for the merge.
You are right, parsing the port is not needed now. It might be in the future if Nyxt gets to guess the right password entry.

**Comment by Ambrevar** [comment]
OK, see https://github.com/atlas-engineer/nyxt/issues/1553 for future discussions.
Closing this then.

**Comment by aartaka** [comment]
Thank you Pierre!

**Comment by Ambrevar** [comment]
Thank you for the initial patch!

**Pull request closed without merging**